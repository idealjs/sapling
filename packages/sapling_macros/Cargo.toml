[package]
name = "sapling_macros"
version = "0.1.0"
edition = "2024"
publish = false

[lib]
proc-macro = true

[dependencies]
quote = { workspace = true }
proc-macro2 = { workspace = true }
syn = { workspace = true }
oxc_allocator = { workspace = true }
oxc_ast = { workspace = true }
oxc_span = { workspace = true }
oxc_syntax = { workspace = true }
oxc_codegen.workspace = true
oxc_parser.workspace = true
oxc_semantic.workspace = true
oxc_traverse.workspace = true
insta.workspace = true
oxc_ast_visit.workspace = true
indextree.workspace = true
