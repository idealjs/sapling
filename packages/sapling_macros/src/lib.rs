use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;
use quote::{ToTokens, quote};
use syn::{Data, DeriveInput, Fields, FieldsNamed, parse_macro_input};

#[proc_macro_attribute]
pub fn tree_builder(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let input = parse_macro_input!(item as DeriveInput);
    let name = &input.ident;
    let vis = &input.vis;
    let attrs = &input.attrs;

    let expanded = match &input.data {
        syn::Data::Struct(s) => match &s.fields {
            Fields::Named(FieldsNamed { named, .. }) => {
                let fields = named.iter();

                quote! {
                    #(#attrs)*
                    #vis struct #name<'a> {
                        #(#fields,)*
                        pub arena: ::indextree::Arena<::oxc_ast::AstKind<'a>>,
                        pub node_stack: ::std::vec::Vec<::indextree::NodeId>,
                        pub allocator: &'a ::oxc_allocator::Allocator,
                        pub scoping: &'a ::oxc_semantic::Scoping,
                    }
                }
            }
            Fields::Unnamed(fields) => {
                let original_fields = fields.unnamed.iter();
                quote! {
                    #(#attrs)*
                    #vis struct #name<'a>(
                        #(#original_fields,)*
                        pub arena: ::indextree::Arena<::oxc_ast::AstKind<'a>>,
                        pub node_stack: ::std::vec::Vec<::indextree::NodeId>,
                        pub allocator: &'a ::oxc_allocator::Allocator,
                        pub scoping: &'a ::oxc_semantic::Scoping,
                    );
                }
            }
            Fields::Unit => {
                quote! {
                    #(#attrs)*
                    #vis struct #name<'a> {
                        pub arena: ::indextree::Arena<::oxc_ast::AstKind<'a>>,
                        pub node_stack: ::std::vec::Vec<::indextree::NodeId>,
                        pub allocator: &'a ::oxc_allocator::Allocator,
                        pub scoping: &'a ::oxc_semantic::Scoping,
                    }
                }
            }
        },
        _ => syn::Error::new(name.span(), "tree_builder can only be applied to structs")
            .to_compile_error(),
    };

    TokenStream::from(expanded)
}

#[proc_macro_attribute]
pub fn tree_builder_mut(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let input = parse_macro_input!(item as DeriveInput);
    let name = &input.ident;
    let vis = &input.vis;
    let attrs = &input.attrs;

    let expanded = match &input.data {
        syn::Data::Struct(s) => match &s.fields {
            Fields::Named(FieldsNamed { named, .. }) => {
                let fields = named.iter();

                quote! {
                    #(#attrs)*
                    #vis struct #name<'a> {
                        #(#fields,)*
                        pub arena: ::indextree::Arena<::oxc_ast::AstType>,
                        pub node_stack: ::std::vec::Vec<::indextree::NodeId>,
                        pub allocator: &'a ::oxc_allocator::Allocator,
                        pub scoping: &'a ::oxc_semantic::Scoping,
                    }
                }
            }
            Fields::Unnamed(fields) => {
                let original_fields = fields.unnamed.iter();
                quote! {
                    #(#attrs)*
                    #vis struct #name<'a>(
                        #(#original_fields,)*
                        pub arena: ::indextree::Arena<::oxc_ast::AstType>,
                        pub node_stack: ::std::vec::Vec<::indextree::NodeId>,
                        pub allocator: &'a ::oxc_allocator::Allocator,
                        pub scoping: &'a ::oxc_semantic::Scoping,
                    );
                }
            }
            Fields::Unit => {
                quote! {
                    #(#attrs)*
                    #vis struct #name<'a> {
                        pub arena: ::indextree::Arena<::oxc_ast::AstType>,
                        pub node_stack: ::std::vec::Vec<::indextree::NodeId>,
                        pub allocator: &'a ::oxc_allocator::Allocator,
                        pub scoping: &'a ::oxc_semantic::Scoping,
                    }
                }
            }
        },
        _ => syn::Error::new(name.span(), "tree_builder can only be applied to structs")
            .to_compile_error(),
    };

    TokenStream::from(expanded)
}



#[proc_macro_derive(TreeBuilder)]
pub fn derive_tree_builder(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    impl_tree_builder(&input)
        .unwrap_or_else(|err| err.to_compile_error())
        .into()
}

fn impl_tree_builder(input: &DeriveInput) -> syn::Result<TokenStream2> {
    let name = &input.ident;

    // Only implement the trait, assuming fields exist
    let expanded = quote! {
        #[automatically_derived]
        impl<'a> ::oxc_ast_visit::Visit<'a> for #name<'a> {
            fn enter_node(&mut self, node: ::oxc_ast::AstKind<'a>) {
                let id = self.arena.new_node(node);
                if let Some(parent) = self.current_parent {
                    parent.append(id, &mut self.arena);
                }
                self.current_parent = Some(id);
            }

            fn leave_node(&mut self, _: ::oxc_ast::AstKind<'a>) {
                if let Some(current) = self.current_parent {
                    if let Some(parent) = current.ancestors(&self.arena).nth(1) {
                        self.current_parent = Some(parent);
                    } else {
                        self.current_parent = None;
                    }
                }
            }
        }
    };

    Ok(expanded)
}
