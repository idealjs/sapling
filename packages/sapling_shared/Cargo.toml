[package]
name = "sapling_shared"
version = "0.1.0"
edition = "2024"
publish = false

[dependencies]
sapling_macros = { path = "../sapling_macros" }

proc-macro2 = { workspace = true }
syn = { workspace = true }
prettyplease = { workspace = true }
cow-utils = { workspace = true }
lazy-regex = { workspace = true }
oxc_allocator.workspace = true
oxc_ast = { workspace = true }
oxc_span = { workspace = true }
oxc_syntax = { workspace = true }
oxc_codegen = { workspace = true }
oxc_parser = { workspace = true }
oxc_transformer = { workspace = true }
oxc_semantic = { workspace = true }
oxc_traverse = { workspace = true }

lazy_static = { workspace = true }
regex = { workspace = true }
phf = { workspace = true, features = ["macros"] }
phf_codegen = { workspace = true }
convert_case = { workspace = true }
indextree.workspace = true
html5ever.workspace = true
markup5ever.workspace = true
markup5ever_rcdom.workspace = true
oxc_ast_visit.workspace = true

[dev-dependencies]
insta = { workspace = true }
