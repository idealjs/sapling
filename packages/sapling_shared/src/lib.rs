pub mod children;
pub mod component;
pub mod condition;
pub mod config;
pub mod dynamic;
pub mod id_gen;
pub mod import;
pub mod jsx_element;
pub mod length_checker;
pub mod native_spread;
pub mod string_utils;
pub mod tag_name;
pub mod text_wrap;
pub mod utils;
pub mod validate;

pub use children::*;
pub use component::*;
pub use condition::*;
pub use config::*;
pub use dynamic::*;
pub use id_gen::*;
pub use import::*;
pub use jsx_element::*;
pub use length_checker::*;
pub use native_spread::*;
pub use string_utils::*;
pub use tag_name::*;
pub use text_wrap::*;
pub use utils::*;

use lazy_static::lazy_static;
use std::collections::HashSet;
use indextree::{Arena, NodeId};
use oxc_allocator::Allocator;
use oxc_ast::{AstKind, AstType};
use oxc_ast::ast::Statement;
use oxc_ast_visit::{Visit, VisitMut};
use oxc_parser::Parser;
use oxc_span::SourceType;
use oxc_traverse::{Traverse, TraverseCtx};

lazy_static! {
    pub static ref RESERVED_NAMESPACES: HashSet<&'static str> = {
        let mut set = HashSet::new();
        set.insert("class");
        set.insert("on");
        set.insert("oncapture");
        set.insert("style");
        set.insert("use");
        set.insert("prop");
        set.insert("attr");
        set.insert("bool");
        set
    };
}

pub trait TreeBuilder<'a>: Visit<'a> {
    fn arena(&self) -> &Arena<AstKind<'a>>;
    fn arena_mut(&mut self) -> &mut Arena<AstKind<'a>>;
    fn node_stack(&self) -> &Vec<NodeId>;
    fn node_stack_mut(&mut self) -> &mut Vec<NodeId>;
    fn current_parent(&self) -> Option<&NodeId> {
        self.node_stack().last()
    }
    fn push_parent(&mut self, node_id: NodeId) {
        self.node_stack_mut().push(node_id);
    }
    fn pop_parent(&mut self) -> Option<NodeId> {
        self.node_stack_mut().pop()
    }
    fn enter_node(&mut self, kind: AstKind<'a>) {
        let node_id = self.arena_mut().new_node(kind);
        if let Some(parent) = self.current_parent() {
            parent.append(node_id, self.arena_mut());
        }
        self.push_parent(node_id);
    }
    fn leave_node(&mut self, _: AstKind<'a>) {
        self.pop_parent();
    }
}

pub trait TreeBuilderMut<'a>: VisitMut<'a> {
    fn arena(&self) -> &Arena<AstType>;
    fn arena_mut(&mut self) -> &mut Arena<AstType>;
    fn node_stack(&self) -> &Vec<NodeId>;
    fn node_stack_mut(&mut self) -> &mut Vec<NodeId>;
    fn current_parent(&self) -> Option<&NodeId> {
        self.node_stack().last()
    }
    fn push_parent(&mut self, node_id: NodeId) {
        self.node_stack_mut().push(node_id);
    }
    fn pop_parent(&mut self) -> Option<NodeId> {
        self.node_stack_mut().pop()
    }
    fn enter_node(&mut self, kind: AstType) {
        let node_id = self.arena_mut().new_node(kind);
        if let Some(parent) = self.current_parent() {
            parent.append(node_id, self.arena_mut());
        }
        self.push_parent(node_id);
    }
    fn leave_node(&mut self, _: AstType) {
        self.pop_parent();
    }
}
