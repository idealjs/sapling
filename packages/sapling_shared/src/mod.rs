//! Sapling JSX transformation modules

// Constants
pub mod html_constants;
pub mod namespace_constants;

// Core JSX handling
pub mod jsx_utils;
pub mod jsx_validator;
pub mod component_transform;
pub mod fragment_transform;
pub mod node_transform;

// Element handling
pub mod dom_element;
pub mod element_attributes;
pub mod element_children;
pub mod element_utils;
pub mod custom_element;

// Attribute handling
pub mod spread_attributes;
pub mod event_handler;

// Dynamic content
pub mod dynamic_wrapper;
pub mod expression_detector;

// Template handling
pub mod template_append;

// String utilities
pub mod string_utils;
pub mod escape_utils;
pub mod text_wrap;
pub mod tag_name;

// Processing stages
pub mod preprocessor;
pub mod post_processor;

// Children handling
pub mod children;
pub mod length_checker;

// Configuration and utilities
pub mod config_utils;
pub mod id_gen;
pub mod native_spread;

// Re-exports of commonly used items
pub use children::{filter_children, check_length};
pub use dom_element::transform_element;
pub use element_attributes::{transform_attributes, set_attr};
pub use jsx_utils::convert_jsx_identifier;
pub use string_utils::{escape_html, trim_whitespace, escape_string_for_template};
pub use tag_name::{jsx_element_name_to_string, tag_name_to_identifier};
