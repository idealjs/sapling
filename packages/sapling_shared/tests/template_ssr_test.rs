use indextree::{Arena, NodeId};
use oxc_allocator::Allocator;
use oxc_allocator::FromIn;
use oxc_ast::ast::Statement;
use oxc_ast::{
    AstKind,
    ast::{Argument, Expression, IdentifierReference, Program, StringLiteral},
};
use oxc_ast_visit::{VisitMut, walk_mut::walk_program};
use oxc_codegen::Codegen;
use oxc_parser::Parser;
use oxc_semantic::{Scoping, Semantic, SemanticBuilder};
use oxc_span::{Atom, SourceType, Span};
use sapling_macros::tree_builder_mut;
use sapling_shared::Config;
use sapling_shared::append_templates_ssr;
use sapling_shared::{
    Template, TreeBuilderMut, create_template_ssr, ssr::template::CreateTemplateInput,
};

#[tree_builder_mut(sapling_shared::TreeBuilderMut<'a>)]
struct TestVisitor<'a> {
    pub config: Config<'a>,
    pub templates: &'a mut Vec<Template<'a>>,
}

impl<'a> VisitMut<'a> for TestVisitor<'a> {
    fn visit_program(&mut self, it: &mut Program<'a>) {
        let expr = create_template_ssr(
            self,
            it,
            &mut CreateTemplateInput {
                template: None,
                exprs: oxc_allocator::Vec::from_array_in(
                    [Expression::Identifier(oxc_allocator::Box::new_in(
                        IdentifierReference {
                            span: Span::default(),
                            name: Atom::from_in("template", self.allocator),
                            reference_id: std::cell::Cell::new(None),
                        },
                        self.allocator,
                    ))],
                    self.allocator,
                ),
                template_values: oxc_allocator::Vec::new_in(self.allocator),
                wont_escape: false,
            },
            "sapling",
        );
        append_templates_ssr(self, it);

        let expr = create_template_ssr(
            self,
            it,
            &mut CreateTemplateInput {
                template: None,
                exprs: oxc_allocator::Vec::from_array_in(
                    [Expression::Identifier(oxc_allocator::Box::new_in(
                        IdentifierReference {
                            span: Span::default(),
                            name: Atom::from_in("template", self.allocator),
                            reference_id: std::cell::Cell::new(None),
                        },
                        self.allocator,
                    ))],
                    self.allocator,
                ),
                template_values: oxc_allocator::Vec::from_array_in(
                    [
                        Expression::StringLiteral(oxc_allocator::Box::new_in(
                            StringLiteral {
                                span: Span::default(),
                                value: Atom::from_in("Hello", self.allocator),
                                raw: None,
                                lone_surrogates: false,
                            },
                            self.allocator,
                        )),
                        Expression::NumericLiteral(oxc_allocator::Box::new_in(
                            oxc_ast::ast::NumericLiteral {
                                span: Span::default(),
                                value: 42.0,
                                raw: None,
                                base: oxc_ast::ast::NumberBase::Decimal,
                            },
                            self.allocator,
                        )),
                        Expression::Identifier(oxc_allocator::Box::new_in(
                            IdentifierReference {
                                span: Span::default(),
                                name: Atom::from_in("world", self.allocator),
                                reference_id: std::cell::Cell::new(None),
                            },
                            self.allocator,
                        )),
                    ],
                    self.allocator,
                ),
                wont_escape: false,
            },
            "sapling",
        );
        it.body
            .push(Statement::ExpressionStatement(oxc_allocator::Box::new_in(
                oxc_ast::ast::ExpressionStatement {
                    span: oxc_span::Span::default(),
                    expression: expr,
                },
                self.allocator,
            )));

        // Test case with template
        let expr = create_template_ssr(
            self,
            it,
            &mut CreateTemplateInput {
                template: Some(sapling_shared::ssr::template::TemplateValue::Single(
                    Atom::from_in("<div>Hello ${}!</div>", self.allocator),
                )),
                exprs: oxc_allocator::Vec::from_array_in(
                    [Expression::Identifier(oxc_allocator::Box::new_in(
                        IdentifierReference {
                            span: Span::default(),
                            name: Atom::from_in("template", self.allocator),
                            reference_id: std::cell::Cell::new(None),
                        },
                        self.allocator,
                    ))],
                    self.allocator,
                ),
                template_values: oxc_allocator::Vec::from_array_in(
                    [Expression::StringLiteral(oxc_allocator::Box::new_in(
                        StringLiteral {
                            span: Span::default(),
                            value: Atom::from_in("world", self.allocator),
                            raw: None,
                            lone_surrogates: false,
                        },
                        self.allocator,
                    ))],
                    self.allocator,
                ),
                wont_escape: false,
            },
            "sapling",
        );
        it.body
            .push(Statement::ExpressionStatement(oxc_allocator::Box::new_in(
                oxc_ast::ast::ExpressionStatement {
                    span: oxc_span::Span::default(),
                    expression: expr,
                },
                self.allocator,
            )));

        // Test case with multiple template parts
        let expr = create_template_ssr(
            self,
            it,
            &mut CreateTemplateInput {
                template: Some(sapling_shared::ssr::template::TemplateValue::Multiple(
                    oxc_allocator::Vec::from_array_in(
                        [
                            Atom::from_in("<div>Hello ", self.allocator),
                            Atom::from_in(", welcome to ", self.allocator),
                            Atom::from_in("!</div>", self.allocator),
                        ],
                        self.allocator,
                    ),
                )),
                exprs: oxc_allocator::Vec::from_array_in(
                    [Expression::Identifier(oxc_allocator::Box::new_in(
                        IdentifierReference {
                            span: Span::default(),
                            name: Atom::from_in("template", self.allocator),
                            reference_id: std::cell::Cell::new(None),
                        },
                        self.allocator,
                    ))],
                    self.allocator,
                ),
                template_values: oxc_allocator::Vec::from_array_in(
                    [
                        Expression::StringLiteral(oxc_allocator::Box::new_in(
                            StringLiteral {
                                span: Span::default(),
                                value: Atom::from_in("user", self.allocator),
                                raw: None,
                                lone_surrogates: false,
                            },
                            self.allocator,
                        )),
                        Expression::StringLiteral(oxc_allocator::Box::new_in(
                            StringLiteral {
                                span: Span::default(),
                                value: Atom::from_in("Sapling", self.allocator),
                                raw: None,
                                lone_surrogates: false,
                            },
                            self.allocator,
                        )),
                    ],
                    self.allocator,
                ),
                wont_escape: false,
            },
            "sapling",
        );
        it.body
            .push(Statement::ExpressionStatement(oxc_allocator::Box::new_in(
                oxc_ast::ast::ExpressionStatement {
                    span: oxc_span::Span::default(),
                    expression: expr,
                },
                self.allocator,
            )));
        append_templates_ssr(self, it);
    }
}

#[test]
fn test_create_template_ssr() {
    let source = "// Initial empty program";
    let allocator = Allocator::default();
    let source_type = SourceType::default().with_typescript(true).with_jsx(true);

    let ret = Parser::new(&allocator, source, source_type).parse();
    let mut program = ret.program;

    let semantic_ret = SemanticBuilder::new().build(&program);
    let mut scoping = semantic_ret.semantic.into_scoping();

    let mut visitor = TestVisitor {
        arena: &mut Arena::new(),
        node_stack: &mut vec![],
        allocator: &allocator,
        scoping: &mut scoping,
        templates: &mut vec![],
        config: Config::default(),
    };
    visitor.visit_program(&mut program);

    // Generate and verify the output code
    let result = Codegen::new().build(&program);
    insta::assert_snapshot!(result.code);
}
