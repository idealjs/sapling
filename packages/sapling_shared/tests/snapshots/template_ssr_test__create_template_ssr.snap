---
source: packages/sapling_shared/tests/template_ssr_test.rs
expression: result.code
---
template;
template;
import { ssr as _$ssr } from "sapling";
_$ssr(template);
_$ssr(template, "user", "Sapling");
// Initial empty program
